src/wallet/wallet.cpp

```diff
@@ -253,6 +253,7 @@ void CWallet::UpgradeKeyMetadata()
     if (m_spk_man) {
         m_spk_man->UpgradeKeyMetadata();
     }
+    SetWalletFlag(WALLET_FLAG_KEY_ORIGIN_METADATA);
```

Comment:
re: https://github.com/bitcoin/bitcoin/pull/17304#discussion_r341791687

> [2590266](https://github.com/bitcoin/bitcoin/commit/2590266c193d27c979ad6287d72555ca3df630bc)
> 
> What if `IsLocked()`?

Nice catch! This seems like a bug. I think either the `m_storage.IsLocked() || m_storage.IsWalletFlagSet(WALLET_FLAG_KEY_ORIGIN_METADATA)` check should move from `LegacyScriptPubKeyMan::UpgradeKeyMetadata` to
`CWallet::UpgradeKeyMetadata`, or preferably, the`SetWalletFlag(WALLET_FLAG_KEY_ORIGIN_METADATA)` call should move back in the other direction.

It's also unclear why `UpgradeKeyMetadata` shoud be a virtual method instead of being specific to the LegacyScriptPubKeyMan class. It doesn't seem like non-legacy keyman types should implement this upgrade function or be accessing this wallet flag.

Reply:
It seems like the `CWallet::UpgradeKeyMetadata()` and `ScriptPubKeyMan::::UpgradeKeyMetadata()` methods could both be dropped and `walletdb.cpp` could call `pwallet->GetLegacyScriptPubKeyMan()` and `LegacyScriptPubKeyMan::UpgradeKeyMetadata` directly for compatibility.

 to do here would be to drop the

Reply:
Changed to be non-virtual and moved that check out to CWallet.